### PR TITLE
Calypso Products: Remove lodash and assertValidProduct

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

This is the first of several PRs to clean up the product identification code that's been moved to the `calypso-products` package by https://github.com/Automattic/wp-calypso/pull/52249 and https://github.com/Automattic/wp-calypso/pull/51955.

Specifically, this PR:

- Removes lodash from the package. All of its uses could be easily handled by native JS.
- Removes `assertValidProduct`. It was being used to validate that objects passed to product identification functions contained a `product_slug`, but most of those functions don't need a `product_slug`, and for those that do, their existing code will handle it.

#### Testing instructions

Automated tests should be sufficient.

### CODE REVIEW

File: packages/calypso-products/src/sort.js [hunk 1]
```diff
@@ -32,23 +33,26 @@ function sortProducts( products ) {
...
-	domainItems = flatten( domainItems );
+	domainItems = domainItems.flat();
 
 	const creditItems = products.filter( isCredits );
 
-	const otherItems = difference( products, planItems, domainItems, includedItems, creditItems );
+	const otherItems = products.filter(
+		( product ) =>
+			! [ ...planItems, ...domainItems, ...includedItems, ...creditItems ].includes( product )
+	);
```
Comment (jsnajdr) [inline packages/calypso-products/src/sort.js]: This version should do fewer array allocations:
```js
const allItems = [ planItems, domainItems, includedItems, creditItems ];
const otherItems = products.filter(
  ( product ) => allItems.every( ( list ) => ! list.includes( product ) )
)
```

File: packages/calypso-products/src/sort.js [hunk 2]
```diff
@@ -32,23 +33,26 @@ function sortProducts( products ) {
...
+	domainItems = products.filter( ( product ) => ! includedItems.includes( product ) );
 	domainItems = domainItems.filter( isDomainProduct );
 	domainItems = Object.entries( groupBy( domainItems, 'meta' ) );
-	domainItems = sortBy( domainItems, function ( pair ) {
+	domainItems = domainItems.sort( function ( pair ) {
 		if ( pair[ 1 ][ 0 ] && pair[ 1 ][ 0 ].cost === 0 ) {
 			return -1;
 		}
 		return pair[ 0 ];
 	} );
```
Comment (jsnajdr) [inline packages/calypso-products/src/sort.js]: Replacing `sortBy` with `Array.prototype.sort` like this won't work. The callbacks are very different:
- `sortBy` callback takes an array item (usually object) as parameter and returns a value (primitive string or number) by which to sort.
- `sort` callback takes two array items as parameters and returns `-1`, `0` or `1` depending on the result of comparison it does.

File: packages/calypso-products/src/sort.js [hunk 3]
```diff
@@ -32,23 +33,26 @@ function sortProducts( products ) {
...
 	domainItems = domainItems.map( function ( pair ) {
-		return sortBy( pair[ 1 ], getDomainProductRanking );
+		return pair[ 1 ].sort( getDomainProductRanking );
 	} );
-	domainItems = flatten( domainItems );
+	domainItems = domainItems.flat();
```
Comment (jsnajdr) [inline packages/calypso-products/src/sort.js]: These two lines could be merged into one `domainItems.flatMap( ( pair ) => ... )`

File: packages/calypso-products/src/sort.js (hunk 2 above)
Comment (jsnajdr) [reply packages/calypso-products/src/sort.js]: Also, beware that `sort` mutates the original array. In this case both `sort` calls are safe.

File: packages/calypso-products/src/utils.js [hunk 4]
```diff
@@ -2,7 +2,7 @@
...
  */
 
-import { map, head, property } from 'lodash';
-
-export const findCurrencyFromPlans = ( plans ) =>
-	head( map( plans, property( 'currency_code' ) ) ) || 'USD';
+export const findCurrencyFromPlans = ( plans ) => {
+	const firstPlanWithCurrency = plans.find( ( plan ) => plan.currency_code );
+	return firstPlanWithCurrency?.currency_code ?? 'USD';
+};
```
Comment (jsnajdr) [inline packages/calypso-products/src/utils.js]: The original function does something different: it looks at the first `plan` in `plans` (never at any other ones) and returns its `plan.currency_code`. If that value is falsy, defaults to `USD`.

File: packages/calypso-products/src/get-domain.js [hunk 5]
```diff
@@ -1,19 +1,12 @@
...
  */
-import { assertValidProduct } from './utils/assert-valid-product';
 import { formatProduct } from './format-product';
 
 export function getDomain( product ) {
 	product = formatProduct( product );
-	assertValidProduct( product );
 
-	const domainToBundle = get( product, 'extra.domain_to_bundle', false );
+	const domainToBundle = product.extra?.domain_to_bundle;
```
Comment (tyxla) [inline packages/calypso-products/src/get-domain.js]: I wonder if we can shorten this to:

```suggestion
	const domainToBundle = product.extra?.domain_to_bundle ?? product.meta;
```

and that way remove these last lines from the function.

File: packages/calypso-products/src/is-jetpack-backup-slug.js [hunk 6]
```diff
@@ -1,9 +1,8 @@
 /**
  * External dependencies
  */
-import { includes } from 'lodash';
 import { JETPACK_BACKUP_PRODUCTS } from './index';
 
 export function isJetpackBackupSlug( productSlug ) {
-	return includes( JETPACK_BACKUP_PRODUCTS, productSlug );
+	return JETPACK_BACKUP_PRODUCTS.includes( productSlug );
```
Comment (tyxla) [inline packages/calypso-products/src/is-jetpack-backup-slug.js:7]: ❤️

File: packages/calypso-products/src/main.js [hunk 7]
```diff
@@ -168,7 +163,7 @@ export function shouldFetchSitePlans( sitePlans, selectedSite ) {
  */
 export function getMonthlyPlanByYearly( planSlug ) {
 	const plan = getPlan( planSlug );
-	if ( has( plan, 'getMonthlySlug' ) ) {
+	if ( plan?.getMonthlySlug ) {
```
Comment (tyxla) [inline packages/calypso-products/src/main.js:166]: This will evaluate the value of `getMonthlySlug`, though, while in the previous version we were only checking if the property exists.

File: packages/calypso-products/src/main.js [hunk 8]
```diff
@@ -183,7 +178,7 @@ export function getMonthlyPlanByYearly( planSlug ) {
  */
 export function getYearlyPlanByMonthly( planSlug ) {
 	const plan = getPlan( planSlug );
-	if ( has( plan, 'getAnnualSlug' ) ) {
+	if ( plan?.getAnnualSlug ) {
```
Comment (tyxla) [inline packages/calypso-products/src/main.js:181]: Same as above with regards to how this check works.

File: packages/calypso-products/src/plans-list.js [hunk 9]
```diff
@@ -11,6 +10,11 @@ import i18n, { translate } from 'i18n-calypso';
 import { isEnabled } from '@automattic/calypso-config';
 import * as constants from './constants';
 
+/* eslint-disable wpcalypso/jsx-classname-namespace */
```
Comment (tyxla) [inline packages/calypso-products/src/plans-list.js]: Can we move this below where it's necessary, and perhaps re-enable it after that line? Or use `eslint-disable-line` instead

File: packages/calypso-products/src/plans-list.js [hunk 10]
```diff
@@ -650,7 +651,7 @@ export const PLANS_LIST = {
 		...getPlanBloggerDetails(),
 		term: constants.TERM_ANNUALLY,
 		getBillingTimeFrame: WPComGetBillingTimeframe,
-		availableFor: ( plan ) => includes( [ constants.PLAN_FREE ], plan ),
+		availableFor: ( plan ) => [ constants.PLAN_FREE ].includes( plan ),
```
Comment (tyxla) [inline packages/calypso-products/src/plans-list.js:655]: Feels weird to use `includes` for an array with a single element, but I guess we can leave it as-is because it wasn't introduced in this PR. Also that way it's consistent with the rest of the `availableFor` definitions.

File: packages/calypso-products/src/sort.js [hunk 11]
```diff
@@ -1,14 +1,15 @@
...
-
 /**
  * Internal dependencies
  */
 import { getDomainProductRanking, isCredits, isDomainProduct, isPlan } from '.';
 
+// https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore
+const groupBy = ( items, key ) =>
+	items.reduce(
+		( r, v, i, a, k = v[ key ] ) => ( ( r[ k ] || ( r[ k ] = [] ) ).push( v ), r ),
```
Comment (tyxla) [inline packages/calypso-products/src/sort.js]: Would be great if we could make this a tad more readable 😉  The part where we declare the array if it doesn't exist may require a bit more eyeballing to read. What do you think?

**Review by tyxla — COMMENTED** [review]
This is great work @sirbrillig 🙌 

I have only a couple of little thoughts to add to the already great review by Jarda.

File: packages/calypso-products/src/main.js (hunk 7 above)
Comment (sirbrillig) [reply packages/calypso-products/src/main.js:166]: True, but since the next line is to call the property as a function, I think that evaluating it is safe here.

File: packages/calypso-products/src/get-domain.js (hunk 5 above)
Comment (sirbrillig) [reply packages/calypso-products/src/get-domain.js]: 👍

File: packages/calypso-products/src/utils.js (hunk 4 above)
Comment (sirbrillig) [reply packages/calypso-products/src/utils.js]: Good catch.

File: packages/calypso-products/src/main.js (hunk 8 above)
Comment (sirbrillig) [reply packages/calypso-products/src/main.js:181]: Same answer. 😁  I think this is safe even if the behavior is different since the next line is `plan.getAnnualSlug()`.

File: packages/calypso-products/src/plans-list.js (hunk 9 above)
Comment (sirbrillig) [reply packages/calypso-products/src/plans-list.js]: Good point. Not sure why that ended up there.

File: packages/calypso-products/src/plans-list.js (hunk 10 above)
Comment (sirbrillig) [reply packages/calypso-products/src/plans-list.js:655]: Yeah, I thought of the same thing but I figured that I'd leave the existing behavior.

**Comment by sirbrillig** [comment]
sort.js will be removed by https://github.com/Automattic/wp-calypso/pull/52211 so I'm going to ignore making further updates to it here. I will rebase this once that has merged.

File: packages/calypso-products/src/sort.js (hunk 11 above)
Comment (sirbrillig) [reply packages/calypso-products/src/sort.js]: Yeah, this was not ideal. I've removed the file, though, so that should resolve it 😂

File: packages/calypso-products/src/sort.js (hunk 3 above)
Comment (sirbrillig) [reply packages/calypso-products/src/sort.js]: Good point. I haven't used `flatMap` before so it didn't occur to me.

File: packages/calypso-products/src/sort.js (hunk 2 above)
Comment (sirbrillig) [reply packages/calypso-products/src/sort.js]: Ah, thank you for catching that. I just saw that the function returned a `-1` sometimes and assumed it worked the same. In any case, this file has now been removed.

File: packages/calypso-products/src/sort.js (hunk 1 above)
Comment (sirbrillig) [reply packages/calypso-products/src/sort.js]: Thanks. That's a good optimization!

File: packages/calypso-products/src/plans-list.js [hunk 12]
```diff
@@ -28,23 +31,24 @@ const getMonthlyTimeframe = () => ( {
...
+	],
 } );
 
+/* eslint-disable wpcalypso/jsx-classname-namespace */
 const plansDescriptionHeadingComponent = {
 	components: {
 		strong: <strong className="plans__features plan-features__targeted-description-heading" />,
 	},
 };
+/* eslint-enable */
```
Comment (tyxla) [inline packages/calypso-products/src/plans-list.js:51]: Did we mean to say `/* eslint-enable wpcalypso/jsx-classname-namespace */` instead? Otherwise, we'd enable any rules we've temporarily disabled, no?

**Review by tyxla — APPROVED** [review]
This looks great from my perspective 👍  Great work @sirbrillig!

File: packages/calypso-products/src/plans-list.js (hunk 12 above)
Comment (sirbrillig) [reply packages/calypso-products/src/plans-list.js:51]: That's a good question. It was actually intentional, as per the recommendation on this internal post p6JqRr-3OA-p2#comment-10916 which I'll quote here:

> if disable is used with a reference to which rules are being disabled, don’t include those rules when adding the enable back in. Why? When someone sees the enable with the ruleset, they wouldn’t know if other rules have been disabled previously or not. When the enable doesn’t have the referenced rule(s), then the reader knows that every disable from earlier in the file, is now re-enabled.

But I don't feel strongly about this if you disagree.
